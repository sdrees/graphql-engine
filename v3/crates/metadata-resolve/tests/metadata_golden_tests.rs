//! Tests that attempt to resolve different metadata files and assert that they parse successfully
//! or fail in the expected way.

use metadata_resolve::MetadataResolveFlagsInternal;

#[test]
fn test_passing_metadata() {
    insta::glob!("passing/**/metadata.json", |path| {
        insta::with_settings!({
            snapshot_path => path.parent().unwrap(),
            snapshot_suffix => "",
            prepend_module_to_snapshot => false,
        }, {
            let metadata_resolve_flags_internal = MetadataResolveFlagsInternal {
                enable_boolean_expression_types: true,
            };

            let metadata_json_text = std::fs::read_to_string(path)
                .unwrap_or_else(|error| panic!("Could not read file {path:?}: {error}"));

            let metadata_json_value = serde_json::from_str(&metadata_json_text)
                .unwrap_or_else(|error| panic!("Could not parse JSON: {error}"));

            let metadata = open_dds::traits::OpenDd::deserialize(metadata_json_value)
                .unwrap_or_else(|error| panic!("Could not deserialize metadata: {error}"));

            let resolved = metadata_resolve::resolve(metadata, metadata_resolve_flags_internal)
                .unwrap_or_else(|error| panic!("Could not resolve metadata: {error}"));

            insta::assert_debug_snapshot!("resolved", resolved);
        });
    });
}

#[test]
fn test_failing_metadata() {
    insta::glob!("failing/**/metadata.json", |path| {
        insta::with_settings!({
            snapshot_path => path.parent().unwrap(),
            snapshot_suffix => "",
            prepend_module_to_snapshot => false,
        }, {
            let metadata_resolve_flags_internal = MetadataResolveFlagsInternal {
                enable_boolean_expression_types: true,
            };

            let metadata_json_text = std::fs::read_to_string(path)
                .unwrap_or_else(|error| panic!("Could not read file {path:?}: {error}"));

            match serde_json::from_str(&metadata_json_text) {
                Ok(metadata_json_value) => {
                    match open_dds::traits::OpenDd::deserialize(metadata_json_value) {
                        Ok(metadata) => {
                            match metadata_resolve::resolve(metadata, metadata_resolve_flags_internal) {
                                Ok(_) => {
                                    panic!("Unexpected success when resolving {path:?}.");
                                }
                                Err(msg) => {
                                    insta::assert_snapshot!("resolve_error", msg);
                                }
                            }
                        }
                        Err(msg) => {
                            insta::assert_snapshot!("deserialize_error", msg);
                        }
                    };
                }

                Err(msg) => {
                    insta::assert_snapshot!("parse_error", msg);
                }
            };
        });
    });
}
