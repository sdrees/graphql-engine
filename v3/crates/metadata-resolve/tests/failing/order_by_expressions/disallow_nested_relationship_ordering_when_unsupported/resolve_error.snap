---
source: crates/metadata-resolve/tests/metadata_golden_tests.rs
expression: msg
input_file: crates/metadata-resolve/tests/failing/order_by_expressions/disallow_nested_relationship_ordering_when_unsupported/metadata.json
---
The following issues were raised but disallowed by compatibility configuration:

The data connector 'myconnector (in subgraph default)' used in model 'mymodel (in subgraph default)' does not support ordering by nested relationships; the orderable relationship 'row_type' from the OrderByExpression 'FooOrderByExpression (in subgraph default)' is nested
